### PR TITLE
T262373: add gallery focus mode

### PR DESCRIPTION
**Pull request by medied**
https://phabricator.wikimedia.org/T262373

The gallery focus mode allows users to hide all other information in the gallery with a single tap to only display the image itself. 

Note this PR builds on top of https://github.com/wikimedia/wikipedia-preview/pull/85, hence the commit list showing all commits. The only commit relevant to focus mode at this point is https://github.com/wikimedia/wikipedia-preview/commit/26d274382c88b8ef88edacd9d706ae3a04180a40. 

Let's aim to merge #85 first.

### CODE REVIEW

**Review by hueitan — APPROVED** [review]
Solution looks good and makes sense to me :+1:

File: src/gallery/slider.js [hunk 1]
```diff
@@ -260,6 +315,30 @@ const clientWidth = window.innerWidth,
 		} )
 	},
 
+	toggleFocusMode = () => {
+		const toggleElements = [
```
Comment (stephanebisson) [inline src/gallery/slider.js:319]: Did you consider only flipping a class here and apply the proper show/hide in CSS?

File: src/gallery/slider.js [hunk 2]
```diff
@@ -79,12 +80,24 @@ const clientWidth = window.innerWidth,
 			},
 
 			author = mediaInfo.author ? mediaInfo.author : msg( lang, 'gallery-unknown-author' ),
-			link = mediaInfo.filePage
+			link = mediaInfo.filePage,
+			description = getImageDescription(),
+
+			isCaptionExpandable = () => {
```
Comment (stephanebisson) [inline src/gallery/slider.js:86]: Why is this "caption expansion" business in this PR?

Comment (medied) [reply src/gallery/slider.js:86]: Yea I built on top of the caption expansion PR given that they overlap a bit in the code, this is what I meant with the note in this PR description

> Note this PR builds on top of #85, hence the commit list showing all commits. The only commit relevant to focus mode at this point is 26d2743.

I gave a heads up in a daily meeting before opening #86

File: src/gallery/slider.js (hunk 1 above)
Comment (medied) [reply src/gallery/slider.js:319]: Seems like a good idea, let me look into that

Comment (medied) [reply src/gallery/slider.js:319]: How about https://github.com/wikimedia/wikipedia-preview/pull/86/commits/ba7d7f7de3895487f2c25148ea003899136a4c5b?

Comment (stephanebisson) [reply src/gallery/slider.js:319]: I think the class should be put higher up on a single element. Which elements should be shown or hidden can be handled in CSS.

**Comment by hueitan** [comment]
I have merged the latest master to this branch, easier for us to review the changes code.

File: src/gallery/slider.js (hunk 1 above)
Comment (medied) [reply src/gallery/slider.js]: Ah I think I see what you meant with your first comment, check out https://github.com/wikimedia/wikipedia-preview/pull/86/commits/3248fae8e6af0117bb5d8324d28f268b1412390a. I like this _much_ better, fully takes advantage of the cascading property of CSS to simplify and: 

* get rid of the nested looping,
* get rid of `focusMode` state variable
* make it more readable and maintainable

Thanks for the great feedback!

**Review by hueitan — APPROVED** [review]
💯 !

**Pull request merged**